### ISSUE TITLE
tf.keras.models.Model's methods get_weights() and set_weights() should retrieve/return Tensors 

### ISSUE THREAD

**Issue by timudk** [tensorflow/tensorflow#29663]
**System information**
- TensorFlow version (you are using): 1.13
- Are you willing to contribute it (Yes/No): Yes

**Describe the feature and the current behavior/state.**
Right now, get_weights() returns a numpy array, and set_weights expects a numpy array as input. Both functions should in my opinion also work for Tensor objects (and therefore be part of the computational graph).

**Will this change the current api? How?**
Two new methods get_weights_tensor(), set_weights_tensor() could be added to tf.keras.models.Models

**Who will benefit with this feature?**
Everybody who wants to modify weights of layers in a loop style.

**Any Other info.**

**Comment by qlzh727** [comment]
"model.weights" is actually the API to retrieve the weight/variable tensor. We don't expect user to reset the weight tensor for a layer/model, which is why we don't have a "set_weights_tensor()" as you proposed.

"Everybody who wants to modify weights of layers in a loop style." is bit confusing to me. Can you provide concrete code snippet to describe what you are trying to achieve?

**Comment by timudk** [comment]
I am using the BFGS optimizer from tensorflow_probablity to train a neural network. This optimizer needs a vector of all variables that need to be updated, i.e. all weights and biases, hence why I would need something like "get_weight_tensor()". If I use "get_weights()" I will get a numpy array which is not part of the computational graph, and therefore I am not able to take derivatives. 

What has worked so far is to define a model in for example this way:
`self.hidden_layer_1 = tf.get_variable('hidden_layer_1', shape=[2, n_neurons], dtype=tf.float64, initializer=tf.initializers.random_normal)
		self.hidden_layer_2 = tf.get_variable('hidden_layer_2', shape=[n_neurons, n_neurons], dtype=tf.float64, initializer=tf.initializers.random_normal)
		self.hidden_layer_3 = tf.get_variable('hidden_layer_3', shape=[n_neurons, n_classes], dtype=tf.float64, initializer=tf.initializers.random_normal)

		self.bias_1 = tf.get_variable('bias_1', [1, n_neurons], dtype=tf.float64, initializer=tf.initializers.random_normal)
		self.bias_2 = tf.get_variable('bias_2', [1, n_neurons], dtype=tf.float64, initializer=tf.initializers.random_normal)
		self.bias_3 = tf.get_variable('bias_3', [1, n_classes], dtype=tf.float64, initializer=tf.initializers.random_normal)

		self.input_var = tf.placeholder(tf.float64, [None, 2]) 

layer = tf.add(tf.matmul(self.input_var, self.hidden_layer_1), self.bias_1)
		layer = tf.nn.relu(layer)
		layer = tf.add(tf.matmul(layer, self.hidden_layer_2), self.bias_2)
		layer = tf.nn.relu(layer)
		layer = tf.add(tf.matmul(layer, self.hidden_layer_3), self.bias_3)`

but i would prefer to use a Sequential model (tf.keras.models.Sequential()) instead.

**Comment by qlzh727** [comment]
As I stated, model.weights is just the get_weight_tensor() method you proposed. It will return you all the weight tensor that is attached to the model/layer. 

From the code snippet you provide, I don't see any part that is swapping the weight of a model. Can you be more explicit about the use case of changing weight tensor?

**Comment by timudk** [comment]
> As I stated, model.weights is just the get_weight_tensor() method you proposed. It will return you all the weight tensor that is attached to the model/layer.

Okay, that's great. How could I set the weights though?

> From the code snippet you provide, I don't see any part that is swapping the weight of a model. Can you be more explicit about the use case of changing weight tensor?

For the above code, it looks like this:
```
def function_evaluation(self, coord):
	split1, split2, split3, split4, split5, split6 = tf.split(coord, [2*self.n_neurons, self.n_neurons, self.n_neurons*self.n_neurons, self.n_neurons, self.n_classes*self.n_neurons, self.n_classes], 0)

	self.hidden_layer_1 = tf.reshape(split1, [2, self.n_neurons])
	self.bias_1 = tf.reshape(split2, [1, self.n_neurons])
	self.hidden_layer_2 = tf.reshape(split3, [self.n_neurons, self.n_neurons])
	self.bias_2 = tf.reshape(split4, [1, self.n_neurons])
	self.hidden_layer_3 = tf.reshape(split5, [self.n_neurons, self.n_classes])
	self.bias_3 = tf.reshape(split6, [1, self.n_classes])

	output = self.model()
	loss = self.loss_function(output)

	hidden_layer_1_gradients = tf.reshape(tf.gradients(loss, self.hidden_layer_1)[0], [2*self.n_neurons, 1])
	bias_1_gradients = tf.reshape(tf.gradients(loss, self.bias_1)[0], [self.n_neurons, 1])
	hidden_layer_2_gradients = tf.reshape(tf.gradients(loss, self.hidden_layer_2)[0], [self.n_neurons*self.n_neurons, 1])
	bias_2_gradients = tf.reshape(tf.gradients(loss, self.bias_2)[0], [self.n_neurons, 1])
	hidden_layer_3_gradients = tf.reshape(tf.gradients(loss, self.hidden_layer_3)[0], [self.n_classes*self.n_neurons, 1])
	bias_3_gradients = tf.reshape(tf.gradients(loss, self.bias_3)[0], [self.n_classes, 1])

	gradients = tf.concat([hidden_layer_1_gradients, bias_1_gradients, hidden_layer_2_gradients, bias_2_gradients, hidden_layer_3_gradients, bias_3_gradients], axis=0)

	return loss, gradients

def training(self, session, x, y):
	result = session.run(tfp.optimizer.bfgs_minimize(self.function_evaluation, initial_position=self.starting_vector, tolerance=self.tol, max_iterations=self.max_itr), feed_dict={self.input_var:x, self.output_var:y})

	return result
```

**Comment by qlzh727** [comment]
OK, I see. Your model have 3 stacking dense layer, but somehow the weights of the dense layer need to be feed in, due the constrain of tfp.optimizer.bfgs_minimize. 

Conceptually, the weight tensor of a layer/model shouldn't changed/swapped since the layer/model is a stateful container. Adding the API to swap the weight tensor will create all sorts of problem and pitfall. If you only need the math calculation part of dense layer, I would suggest to just stay with your current code snippet.

@fchollet, do you have any other suggestions?

**Comment by qlzh727** [comment]
Checked with Francois offline yesterday. He also agrees that allowing user to change the weight tensor will cause big issue and confuse user. We will close this issue as we don't intend to address it. You can leave your current code as is, or create a custom layer.

**Comment by bionicles** [comment]
@qlzh727 @fchollet @timudk 

hi, this makes it seemingly impossible to build a hypernetwork wrapper which uses "get_weights" and "set_weights" -- this API exists, but just doesnt work inside tf 2.0 bc we need numpy arrays to set tensorflow keras layer weights (which seems stupid, why don't we just use tensors?)

I get that it "might cause issues" but ... the API as-is contains a broken "set_weights" feature, which fails in @tf.function decorated call methods, so, please either fix set_weights to use tensors or delete set_weights!

**Comment by MedAbdelkaderAbba** [comment]
I have the same problem with set_weights, when using tensorflow_probability.

**Comment by fachu000** [comment]
@qlzh727 @fchollet @timudk

In popular metalearning approaches such as MAML, one needs to take gradients w.r.t. the initializations of the weights of the layers. 

It seems that people so far are not using Keras model/layer for metalearning because of the API does not allow it. See e.g. https://github.com/cbfinn/maml/blob/master/maml.py#L202, https://github.com/dragen1860/MAML-TensorFlow/blob/master/maml.py, https://github.com/sudharsan13296/Hands-On-Meta-Learning-With-Python/tree/master/06.%20MAML%20and%20it's%20Variants .

In my understanding, addressing this issue would allow metalearning with Keras.
